### ISSUE TITLE
10.38 - Jit bug with ungrouped OR

### ISSUE THREAD

**Issue by SjonHortensius** [PhilipHazel/pcre2#21]
with JIT enabled, https://github.com/PhilipHazel/pcre2/commit/dc5f96663597572f694147aeec3525003c351123 triggers a bug:

re `/(?P<size>\d+)m|M/` with data `4M` will not match, while it should return 'M' (which it does without JIT)

**Comment by PhilipHazel** [comment]
I can confirm this bug. Pity it wasn't discovered during the month of testing 10.38-RC1.

**Comment by SjonHortensius** [comment]
yeah, sorry. My primary involvement is with PHP and I only found this because they [included](https://github.com/php/php-src/commit/a2471383fec332ae30827c7e3f4f9451420f1f0b) that change which caused [differences in behaviour](https://3v4l.org/bughunt/8.1rc2/8.1rc1+8.0.10) among the 4M scripts I'm able to test

**Comment by fghamsary** [comment]
I confirm that we have the same problem in while using docopt on PHP, until update this morning on libpcre2 many of my applications stopped working, and finally we've arrived at the same conclusion in our company.
Hope that this will be fixed ASAP.
PS: I'm using Ubuntu 20.04 and PHP 7.4.24
With libpcre2 version: 10.37

**Comment by SjonHortensius** [comment]
@fghamsary PHP ships with pcre2 bundled - and this change was only included in PHP 8.1 rc2 and rc3, not in 7.4.24 (which works [fine)](https://3v4l.org/slE13). Also - pcre2 10.37 does not include this change either. Are you sure you run into this bug and not something else?

**Comment by fghamsary** [comment]
> @fghamsary PHP ships with pcre2 bundled - and this change was only included in PHP 8.1 rc2 and rc3, not in 7.4.24 (which works [fine)](https://3v4l.org/slE13). Also - pcre2 10.37 does not include this change either. Are you sure you run into this bug and not something else?

Here is my observation, tell me if I'm wrong.
If I execute the following code:
`preg_split('@\s+|(\S*<.*>)@', 'test all [ test<t>', null, PREG_SPLIT_DELIM_CAPTURE)`
The desired result should be something like this:
`["test", "all", "[", "", "test<t>", ""]`
Which is what I get on php 7.4.24 on Ubuntu 20.04 and libpcre2: 10.36.
But if I run exactly the same code with php 7.4.24 on Ubuntu 20.04 and libpcre2: 10.37 which I mentioned before, the result will be this:
`["test ", "all [ ", "test<t>", ""]`
This is just an example from simplified code on docopt code.

**Comment by fghamsary** [comment]
And more over even to confirm that I've disabled JIT on the php.ini and the problem is resolved! So I'm pretty sure that we are talking about the same problem.

**Comment by fghamsary** [comment]
If you need any more information about any part of the configuration or the code, I'll be glad to help you find the problem and resolve it, but as it was really the same issue for me I thought that it could be related. But maybe there is another things as you said that we can miss.
Just another precision I use the http://ppa.launchpad.net/ondrej/php/ubuntu package version of php on Ubuntu.
Maybe something wrong with that, but I doubt it as until yesterday everything was working fine, and just today after the update of libpcre2 on my Ubuntu I encountered the problem. No update of php has been done. So, please do let me know.

**Comment by SjonHortensius** [comment]
@fghamsary you might be using a distro that replaces php's pcre2, but it's not really relevant to this issue as @PhilipHazel already confirmed the bug, which has nothing to do with PHP other then that was how I found it

**Comment by fghamsary** [comment]
Thanks for the reactivity, my problem was on https://github.com/oerdnj/deb.sury.org/issues/1653 this one. Sorry for the comments, and thanks a lot :)

**Comment by zherczeg** [comment]
I am not sure reverting the original bug is a good idea. Instead the #22 should fix this. I recommend to test release candidates to detect such issues before release. Please understand that I work on this project in my free time, so I cannot do things instantly.

**Comment by cmb69** [comment]
>  Please understand that I work on this project in my free time, […]

What is highly appreciated! :)

Note that PHP had backported the fix to the bundled PCRE 10.35/10.37, and so we reverted immediately before the release of the new RCs; a performace regression is less worse than a functional regression. It's still possible that both patches will be backported for the next PHP versions.

Thank you for the quick fix!

**Comment by ltrzesniewski** [comment]
To be fair, dc5f96663597572f694147aeec3525003c351123 wasn't included in 10.38-RC1, so the problem couldn't have been detected that way.

@PhilipHazel do you think releasing a patch version like 10.38.1 would be appropriate, or will the fix only be available in the next release cycle?

(I maintain a [.NET library](https://github.com/ltrzesniewski/pcre-net) that wraps PCRE2, and I'm not sure if I should release a version that includes 10.38 + #22 or if I should wait for a patch version instead).

**Comment by zherczeg** [comment]
I think the release cycle should be improved in some way. Before releasing 10.38, we have waited for a month for feedback, but none came. If we release another version, and people will come again that there are bugs, then we need to do another release, and I don't really like that. I am not against a 10.38.1, but there should be a broader testing before it goes live. Perhaps we need to figure out something to request some testing from the community before releases first.

**Comment by PhilipHazel** [comment]
I agree, but I am not sure how to improve things. One of the problems, I suppose, is that a lot of users don't use PCRE2 directly, but via applications such as PHP, so unless those applications put out matching testing versions, no testing gets done. Maybe we should just put out releases more frequently?

**Comment by nikic** [comment]
This seems like the kind of problem a fuzzer should be able to find relatively easily, but running a jit and non-jit match and comparing that they are the same (if no limit has been hit). I took a look at https://github.com/PhilipHazel/pcre2/blob/master/src/pcre2_fuzzsupport.c, and it doesn't seem to do something like that right now, if I'm reading the code right. It performs matches two times with and without flags, but it doesn't try to check that the output is the same for flags that shouldn't affect results.

(I'm doing something along these lines with PHP's JIT compiler right now.)

**Comment by SjonHortensius** [comment]
@MatthewVernon while https://3v4l.org supplies me with an ever growing set of testcases - extracting both the regex and expected result is near impossible to automate and I expect there to be a lot of noise as well.

It seems to me a fuzzer would make a lot of sense for pce2 to adopt as testing strategy
